hbase-server/src/main/java/org/apache/hadoop/hbase/io/FSDataInputStreamWrapper.java

```diff
@@ -270,39 +267,23 @@ public void unbuffer() {
       if (this.instanceOfCanUnbuffer == null) {
         // To ensure we compute whether the stream is instance of CanUnbuffer only once.
         this.instanceOfCanUnbuffer = false;
-        Class<?>[] streamInterfaces = streamClass.getInterfaces();
-        for (Class c : streamInterfaces) {
-          if (c.getCanonicalName().toString().equals("org.apache.hadoop.fs.CanUnbuffer")) {
-            try {
-              this.unbuffer = streamClass.getDeclaredMethod("unbuffer");
-            } catch (NoSuchMethodException | SecurityException e) {
-              if (isLogTraceEnabled) {
-                LOG.trace("Failed to find 'unbuffer' method in class " + streamClass
-                    + " . So there may be a TCP socket connection "
-                    + "left open in CLOSE_WAIT state.", e);
-              }
-              return;
-            }
-            this.instanceOfCanUnbuffer = true;
-            break;
-          }
+        if (wrappedStream instanceof CanUnbuffer) {
+          this.unbuffer = (CanUnbuffer) wrappedStream;
+          this.instanceOfCanUnbuffer = true;
         }
       }
       if (this.instanceOfCanUnbuffer) {
```

Comment:
If `stream` is ever null, this will throw an NPE because `instanceOfCanUnbuffer` is still null.

Reply:
@joshelser Thanks for the review! 
`
if (this.instanceOfCanUnbuffer == null) {
        // To ensure we compute whether the stream is instance of CanUnbuffer only once.
        this.instanceOfCanUnbuffer = false;
`
in above code, already handle the instanceOfCanUnbuffer == null.

Reply:
If `stream` is null, you never get to your `instanceOfCanUnbuffer == null` check.

Reply:
@joshelser  If `stream` is null, it will do nothing in the `unbuffer()` method, all the operations in the `if (stream != null)` judgment,  and the `instanceOfCanUnbuffer ` only used in `unbuffer()` ,  so it will NOT hit the NPE.

Reply:
I'm really struggling because I don't understand why you don't see the same issue I am seeing. Let me try to be super clear. Sorry if I'm missing something obvious.
1. We start this method `unbuffer()` and `instanceOfCanUnbuffer = null`
2. L260, `getStream()` returns `null` (whether or not this is feasible in the code, let's play along)
3. We immediately go to L275 where `instanceOfCanUnbuffer == null` and we throw an NPE.
3. L261, the `if` block fails, and we do not take it as `stream == null`

Is this code-path flawed?

Reply:
@joshelser  Take easy!  And Thanks for your patience! I totally understand your point. Let me show what I see,
1. We start this method unbuffer() and instanceOfCanUnbuffer = null
2. L260, if `getStream()` returns null 
3. It will go to L289
![image](https://user-images.githubusercontent.com/12442769/84341470-c2096200-abd5-11ea-88f1-5bf5747459a1.png)

Reply:
Sorry, the code-folding got me, thanks for the image!